### PR TITLE
Implement OAuth 2.0 for WebFlux

### PR DESCRIPTION
**Pull request by mraible**
Still need to implement the following:

* [x] Make authorities extractor work (might require upgrading to Spring Security 5.2)
* [x] Make User sync code in `UserService` non-blocking
* [x] Configure CORS
* [x] Save new roles that aren't in the database
* [x] Fix Logout
* [x] Enable AuditResourceIT#getAudit() and get it to pass

### CODE REVIEW

**Comment by mraible** [comment]
I tried to fix the `syncUserWithIdP()` method today to make it non-blocking, but failed. Here's my attempt:

```java
private User syncUserWithIdP(Map<String, Object> details, User user) {
    // save authorities in to sync user roles/groups between IdP and JHipster's local database
    Flux<String> userAuthorities = Flux.fromStream(user.getAuthorities().stream().map(Authority::getName));
    Flux<String> newAuthorities = Flux.merge(userAuthorities, getAuthorities()).distinct();

    // check to see if user's authorities exist in database, add them if they don't
    // todo: confirm that this saves new authorities from the user that's passed in
    newAuthorities.flatMap(authority -> {
        log.debug("Saving authority '{}' in local database", authority);
        Authority authorityToSave = new Authority();
        authorityToSave.setName(authority);
        return authorityRepository.save(authorityToSave);
    }).subscribe();

    // save account in to sync users between IdP and JHipster's local database
    return userRepository.findOneByLogin(user.getLogin())
        .switchIfEmpty(userRepository.save(user).thenReturn(user))
        .flatMap(existingUser -> {
            // if IdP sends last updated information, use it to determine if an update should happen
            if (details.get("updated_at") != null) {
                Instant dbModifiedDate = existingUser.getLastModifiedDate();
                Instant idpModifiedDate = new Date(Long.valueOf((Integer) details.get("updated_at"))).toInstant();
                if (idpModifiedDate.isAfter(dbModifiedDate)) {
                    log.debug("Updating user '{}' in local database", user.getLogin());
                    return updateUser(user.getFirstName(), user.getLastName(), user.getEmail(),
                        user.getLangKey(), user.getImageUrl()).then();
                }
                // no last updated info, blindly update
            } else {
                log.debug("Updating user '{}' in local database", user.getLogin());
                return updateUser(user.getFirstName(), user.getLastName(), user.getEmail(),
                    user.getLangKey(), user.getImageUrl()).then();
            }
        });
}
```

Issues:

* The last `flatMap()` expects a return statement, so it doesn't compile.
* The `updateUser()` method now returns a `Mono<User>`, so I'm not sure if this method should return a `Mono<User>`. If it should, how does the `getUserFromAuthentication()` method need to change to support it? Currently, it has the following as a return statement:

```java
return Mono.just(new UserDTO(syncUserWithIdP(attributes, user)));
```

@cbornet: any ideas?

**Comment by mraible** [comment]
I was able to figure out how to do user authorities mapping (thanks to https://github.com/spring-projects/spring-security/issues/6266#issuecomment-569066606), replacing the `GrantedAuthoritiesMapper` bean with the following:

```java
@Bean
public ReactiveOAuth2UserService<OidcUserRequest, OidcUser> oidcUserService() {
    final OidcReactiveOAuth2UserService delegate = new OidcReactiveOAuth2UserService();

    return (userRequest) -> {
        // Delegate to the default implementation for loading a user
        return delegate.loadUser(userRequest).map(user -> {
            Set<GrantedAuthority> mappedAuthorities = new HashSet<>();

            user.getAuthorities().forEach(authority -> {
                if (authority instanceof OidcUserAuthority) {
                    OidcUserAuthority oidcUserAuthority = (OidcUserAuthority) authority;
                    mappedAuthorities.addAll(SecurityUtils.extractAuthorityFromClaims(oidcUserAuthority.getUserInfo().getClaims()));
                }
            });

            return new DefaultOidcUser(mappedAuthorities, user.getIdToken(), user.getUserInfo());
        });
    };
}
```

I also fixed `UserService` so it compiles. This seems to work since the following is printed:

```
2019-12-30 13:08:50.998 DEBUG 69686 --- [ctor-http-nio-3] c.m.myapp.aop.logging.LoggingAspect      : Enter: com.mycompany.myapp.service.UserService.getUserFromAuthentication() with argument[s] = [org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken@4f0ad11: Principal: Name: [4c973896-5761-41fc-8217-07c5d13a004b], Granted Authorities: [[ROLE_ADMIN, ROLE_USER]], ...]
```

However, I don't think this logic is correct since the user's authorities get saved everything time.

```java
Flux<String> newAuthorities = Flux.merge(userAuthorities, getAuthorities()).distinct();

// check to see if user's authorities exist in database, add them if they don't
newAuthorities
    .filter(authority -> !authority.startsWith("SCOPE_"))
    .flatMap(authority -> {
        log.debug("Saving authority '{}' in local database", authority);
        Authority authorityToSave = new Authority();
        authorityToSave.setName(authority);
        return authorityRepository.save(authorityToSave);
    }).subscribe();
```

**Comment by mraible** [comment]
@cbornet If you have time this week, I'd love your assistance on this one. See this PR's description for issues that aren't working. You can reproduce them with the following steps:

```
git clone https://github.com/jhipster/jhipster.git && cd jhipster && mvn install
cd ..
git clone -b oauth2 git@github.com:mraible/jhipster-reactive-monolith-oauth2.git reactive-mono
cd reactive-mono
docker-compose -f src/main/docker/keycloak.yml up -d
./mvnw
```

Log in and logout. You'll see the logout URL is not resolved. Calling [toString() on the Mono](https://github.com/mraible/jhipster-reactive-monolith-oauth2/blob/oauth2/src/main/java/com/mycompany/myapp/web/rest/LogoutResource.java#L40) doesn't work.

For the `AuditResourceIT#getAudit()`, you'll need to remove `@Disabled`. I'm not sure why this doesn't work. It looks like the data is being inserted before it's edited.

Finally, the logic to compare existing authorities in the database against new ones in the user's details doesn't seem to be right.

**Comment by cbornet** [comment]
Sure I'll have a look asap.

File: src/main/java/com/mycompany/myapp/service/UserService.java [hunk 1]
```diff
@@ -143,48 +146,49 @@ public UserService(UserRepository userRepository, AuthorityRepository authorityR
...
+            .filter(authority -> !authority.startsWith("SCOPE_"))
+            .flatMap(authority -> {
                 log.debug("Saving authority '{}' in local database", authority);
                 Authority authorityToSave = new Authority();
                 authorityToSave.setName(authority);
-                authorityRepository.save(authorityToSave);
-            }
-        }
+                return authorityRepository.save(authorityToSave);
+            }).subscribe();
```
Comment (cbornet) [inline src/main/java/com/mycompany/myapp/service/UserService.java]: You should never have to call `subscribe` in a WebFlux project. Subscription is done internally by Spring and the methods should all be chained together to propagate the backpressure.

**Comment by cbornet** [comment]
> For the AuditResourceIT#getAudit(), you'll need to remove @Disabled. I'm not sure why this doesn't work. It looks like the data is being inserted before it's edited.

It's weird indeed. The code in my PR https://github.com/jhipster/generator-jhipster/pull/11042/files is very similar and works OK when tested with Webflux+mongo.
Anyway, I don't think the equivalent of AuditEventRepository exists for Webflux so I had to do the recording of security events manually for JWT and session. Do you know if you can intercept successful/failed auth and save an audit event with reactive OAuth2 ? If it's not possible, maybe the auditing code should be removed for now

**Comment by cbornet** [comment]
For the logout, here is code that I think works better
```java
    public Mono<Map<String, String>> logout(@AuthenticationPrincipal(expression = "idToken") OidcIdToken idToken) {
        return this.registration
            .map(oidc -> oidc.getProviderDetails()
                .getConfigurationMetadata()
                .get("end_session_endpoint")
                .toString()
            )
            .map(logoutUrl -> {
                Map<String, String> logoutDetails = new HashMap<>();
                logoutDetails.put("logoutUrl", logoutUrl);
                logoutDetails.put("idToken", idToken.getTokenValue());
                return logoutDetails;
            });
    }
```
However something else must be wrong as I am still logged in after clicking the sign out button.

**Comment by cbornet** [comment]
For the syncUserWithIdP, can you test the following code ?
```java
    private Mono<User> syncUserWithIdP(Map<String, Object> details, User user) {
        // save authorities in to sync user roles/groups between IdP and JHipster's local database
        Collection<String> userAuthorities =
            user.getAuthorities().stream().map(Authority::getName).collect(Collectors.toList());

        return getAuthorities().collectList()
            .flatMapMany(dbAuthorities -> {
                List<Authority> authoritiesToSave = userAuthorities.stream()
                    .filter(authority -> !dbAuthorities.contains(authority))
                    .map(authority -> {
                        Authority authorityToSave = new Authority();
                        authorityToSave.setName(authority);
                        return authorityToSave;
                    })
                    .collect(Collectors.toList());
                return Flux.fromIterable(authoritiesToSave);
            })
            .doOnNext(authority -> log.debug("Saving authority '{}' in local database", authority))
            .flatMap(authorityRepository::save)
            .then(userRepository.findOneByLogin(user.getLogin()))
            .switchIfEmpty(userRepository.save(user))
            .flatMap(existingUser -> {
                // if IdP sends last updated information, use it to determine if an update should happen
                if (details.get("updated_at") != null) {
                    Instant dbModifiedDate = existingUser.getLastModifiedDate();
                    Instant idpModifiedDate = new Date(Long.valueOf((Integer) details.get("updated_at"))).toInstant();
                    if (idpModifiedDate.isAfter(dbModifiedDate)) {
                        log.debug("Updating user '{}' in local database", user.getLogin());
                        return updateUser(user.getFirstName(), user.getLastName(), user.getEmail(),
                            user.getLangKey(), user.getImageUrl());
                    }
                    // no last updated info, blindly update
                } else {
                    log.debug("Updating user '{}' in local database", user.getLogin());
                    return updateUser(user.getFirstName(), user.getLastName(), user.getEmail(),
                        user.getLangKey(), user.getImageUrl());
                }
                return Mono.empty();
            })
            .thenReturn(user);
    }
```

**Comment by mraible** [comment]
@cbornet I tried your logout code. It doesn't seem to return a response when there's a POST to the `/api/logout` endpoint. It should return the ID token and the logout endpoint. It used to return only the ID token, now it returns neither. Also, the test fails:

```
java.lang.AssertionError: JSON path "$.logoutUrl" expected:<MonoMapFuseable> but was:<https://jhipster.org/logout>
Expected :MonoMapFuseable
Actual   :https://jhipster.org/logout
```

**Comment by cbornet** [comment]
Is this from the test ? It looks like the expected is wrong, not the actual !

**Comment by mraible** [comment]
> Is this from the test ? It looks like the expected is wrong, not the actual !

Yes. Do you know how the test should be updated?

It looks like the following is not working with Keycloak, maybe it needs the ID Token like Okta does?

```ts
logout(): void {
  this.authServerProvider.logout().subscribe((logout: Logout) => {
    let logoutUrl = logout.logoutUrl;
    const redirectUri = `${location.origin}${this.location.prepareExternalUrl('/')}`;

    // if Keycloak, uri has protocol/openid-connect/token
    if (logoutUrl.includes('/protocol')) {
      logoutUrl = logoutUrl + '?redirect_uri=' + redirectUri;
    } else {
      // Okta
      logoutUrl = logoutUrl + '?id_token_hint=' + logout.idToken + '&post_logout_redirect_uri=' + redirectUri;
    }
    window.location.href = logoutUrl;
  });
}
```

**Comment by cbornet** [comment]
For the test I would do
```java
String logoutUrl = this.registrations.findByRegistrationId("oidc")
            .map(oidc -> oidc.getProviderDetails().getConfigurationMetadata()
                .get("end_session_endpoint").toString())
            .block();
```
Are you sure the logout endpoint doesn't work for you ? It works on my machine :smiling_imp:

**Comment by mraible** [comment]
@cbornet Thanks. I was able to fix the `LogoutResourceIT`. Unfortunately, logout in Chrome still doesn't work.

**Comment by mraible** [comment]
I tested logout with Okta and it fails there too, proving that it's not a parameter issue. I can see the ID token being sent successfully to Okta.
```
¯\_(ツ)_/¯
```

**Comment by mraible** [comment]
@cbornet After updating the code to use your `syncUserWithIdP()` method, the `AccountResourceIT` is [failing](https://github.com/mraible/jhipster-reactive-monolith-oauth2/pull/1/checks?check_run_id=378865622#step:5:175). Any ideas?

```
{"app_name":"jhipster","app_port":"-1","level":"ERROR","logger_name":"o.z.p.s.c.AdviceTraits",
  "message":"Internal Server Error","stack_trace":"java.lang.NullPointerException: 
  The mapper returned a null value.\n\tat java.util.Objects.requireNonNull(Objects.java:246)\n\t
  Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: \nError has been 
  observed at the following site(s):\n\t|_ checkpoint ⇢ Handler 
  com.mycompany.myapp.web.rest.AccountResource#getAccount(Principal) 
  [DispatcherHandler]\nStack trace:\n\t\tat 
```

Related: any ideas on how to make the stack traces easier to read?

For the logout issue, I [posted a question on Stack Overflow](https://stackoverflow.com/questions/59650868/how-do-i-logout-of-when-using-spring-webflux-and-oauth-2-0-login).

**Comment by cbornet** [comment]
For the stack trace, there's a parameter in application.yml to format as json. I also think it's not great to have structured logs by default for the tests...

**Comment by cbornet** [comment]
The 500 error in AccountResourceIT is because a `null` is returned in a map step of `SecurityUtils::getCurrentUserLogin` which is strictly forbidden (I wrote that code, shame on me).
It can be fixed by wrapping it in a `Mono` or an `Optional`.
Here is an example with a Mono wrapping:

```java
    public static Mono<String> getCurrentUserLogin() {
        return ReactiveSecurityContextHolder.getContext()
            .map(SecurityContext::getAuthentication)
            .flatMap(authentication -> Mono.justOrEmpty(extractPrincipal(authentication)));
    }

    private static String extractPrincipal(Authentication authentication) {
        if (authentication.getPrincipal() instanceof UserDetails) {
            UserDetails springSecurityUser = (UserDetails) authentication.getPrincipal();
            return springSecurityUser.getUsername();
        } else if (authentication instanceof JwtAuthenticationToken) {
            return (String) ((JwtAuthenticationToken)authentication).getToken().getClaims().get("preferred_username");
        } else if (authentication.getPrincipal() instanceof DefaultOidcUser) {
            Map<String, Object> attributes = ((DefaultOidcUser) authentication.getPrincipal()).getAttributes();
            if (attributes.containsKey("preferred_username")) {
                return (String) attributes.get("preferred_username");
            }
        } else if (authentication.getPrincipal() instanceof String) {
            return (String) authentication.getPrincipal();
        }
        return null;
    }
```

**Comment by mraible** [comment]
I figured out how to do logout. You need to get a `WebSession` and call `invalidate()` on it. https://stackoverflow.com/a/59705972/65681

**Comment by mraible** [comment]
@cbornet I added success and failure logging for `oauth2Login()` in https://github.com/mraible/jhipster-reactive-monolith-oauth2/pull/1/commits/6c263776a5207652d64c4784ec1ae16cb5a6b643. However, I don't see how this would affect `AuditResourceIT#getAudit()` since it's a simple CRUD operation. It seems to just add a record to the database and then try to retrieve it from the resource/service by ID.

```java
@Test
@Disabled
public void getAudit() {
    // Initialize the database
    auditEventRepository.save(auditEvent).block();

    // todo: figure out why id is not retrievable in AuditResource::get
    // Get the audit
    webTestClient.get().uri("/management/audits/{id}", auditEvent.getId())
        .exchange()
        .expectStatus().isOk()
        .expectHeader().contentType(MediaType.APPLICATION_JSON_VALUE)
        .expectBody()
        .jsonPath("$.principal").value(hasItem(SAMPLE_PRINCIPAL));
}
```

Is there something more involved that causes the event to not be saved to the database?

**Comment by mraible** [comment]
Nevermind, I had to remove success/failure from the security config because it was blocking authentication. I'll try merging generator-jhipster's `master` branch.

**Comment by cbornet** [comment]
Ok. To call invalidate, don't use subscribe. It would break the reactive streams chain. Subscribe will be called by the Webflux dispatcher. So instead, call invalidate in the chain. For instance, with `then` or `flatmap`.

**Comment by cbornet** [comment]
You can write like this for instance
```java
    public Mono<Map<String, String>> logout(@AuthenticationPrincipal(expression = "idToken") OidcIdToken idToken, WebSession session) {
        return session.invalidate().then(
            this.registration.map(oidc -> oidc.getProviderDetails().getConfigurationMetadata().get("end_session_endpoint").toString())
                .map(logoutUrl -> {
                    Map<String, String> logoutDetails = new HashMap<>();
                    logoutDetails.put("logoutUrl", logoutUrl);
                    logoutDetails.put("idToken", idToken.getTokenValue());
                    return logoutDetails;
                })
        );
    }
```

**Comment by mraible** [comment]
I was able to get an authentication success handler to work, but I can't figure out how to get it to redirect back to the default location.

```java
private Mono<Void> onAuthenticationSuccess(WebFilterExchange exchange, Authentication authentication) {
    return Mono.just(authentication.getPrincipal())
        .filter(principal -> principal instanceof OidcUser)
        .map(principal -> ((OidcUser) principal).getPreferredUsername())
        .filter(login -> !Constants.ANONYMOUS_USER.equals(login))
        .flatMap(auditEventService::saveAuthenticationSuccess)
        .then();
}
````

The default `RedirectServerAuthenticationSuccessHandler` has:

```java
@Override
public Mono<Void> onAuthenticationSuccess(WebFilterExchange webFilterExchange,
	Authentication authentication) {
	ServerWebExchange exchange = webFilterExchange.getExchange();
	return this.requestCache.getRedirectUri(exchange)
		.defaultIfEmpty(this.location)
		.flatMap(location -> this.redirectStrategy.sendRedirect(exchange, location));
}
```

I'm guessing we just need to figure out a way to chain the two together.

**Comment by cbornet** [comment]
> I'm guessing we just need to figure out a way to chain the two together.

Maybe just put the `RedirectServerAuthenticationSuccessHandler` code in the `then`?

**Comment by mraible** [comment]
> Maybe just put the `RedirectServerAuthenticationSuccessHandler` code in the then?

Yeah, I thought of that, but was never able to chain them together successfully. I timeboxed my attempt to an hour.

**Comment by cbornet** [comment]
Can you commit your code for the success handler ? I'll have a look.

**Comment by mraible** [comment]
@cbornet Added in https://github.com/mraible/jhipster-reactive-monolith-oauth2/pull/1/commits/9dd81cbb834f51f19f7748fc48f4a66375c41a18.

**Comment by cbornet** [comment]
I tried this which seems to work

```java
    private ServerAuthenticationSuccessHandler redirectServerAuthenticationSuccessHandler = new RedirectServerAuthenticationSuccessHandler();

    private Mono<Void> onAuthenticationSuccess(WebFilterExchange exchange, Authentication authentication) {
        return redirectServerAuthenticationSuccessHandler.onAuthenticationSuccess(exchange, authentication)
            .thenReturn(authentication.getPrincipal())
            .filter(principal -> principal instanceof OidcUser)
            .map(principal -> ((OidcUser) principal).getPreferredUsername())
            .filter(login -> !Constants.ANONYMOUS_USER.equals(login))
            .flatMap(auditEventService::saveAuthenticationSuccess)
            .then();
    }
```

**Pull request closed without merging**